streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java

```diff
@@ -185,32 +200,16 @@ public void registerStore(final StateStore store, final StateRestoreCallback sta
 
         log.info("Restoring state for global store {}", store.name());
         final List<TopicPartition> topicPartitions = topicPartitionsForStore(store);
-        Map<TopicPartition, Long> highWatermarks = null;
 
-        int attempts = 0;
-        while (highWatermarks == null) {
-            try {
-                highWatermarks = globalConsumer.endOffsets(topicPartitions);
-            } catch (final TimeoutException retryableException) {
-                if (++attempts > retries) {
-                    log.error("Failed to get end offsets for topic partitions of global store {} after {} retry attempts. " +
-                        "You can increase the number of retries via configuration parameter `retries`.",
-                        store.name(),
-                        retries,
-                        retryableException);
-                    throw new StreamsException(String.format("Failed to get end offsets for topic partitions of global store %s after %d retry attempts. " +
-                            "You can increase the number of retries via configuration parameter `retries`.", store.name(), retries),
-                        retryableException);
-                }
-                log.debug("Failed to get end offsets for partitions {}, backing off for {} ms to retry (attempt {} of {})",
-                    topicPartitions,
-                    retryBackoffMs,
-                    attempts,
-                    retries,
-                    retryableException);
-                Utils.sleep(retryBackoffMs);
-            }
-        }
+        final Map<TopicPartition, Long> highWatermarks = new HashMap<>();
+        retryUntilSuccessOrThrowOnTaskTimeout(
+            () -> highWatermarks.putAll(globalConsumer.endOffsets(topicPartitions)),
```

Comment:
No need to change anything, just a note: It's mildly concerning to see side-effecting operations inside a retry loop, since there's no guarantee that failed attempts won't leave any garbage in the system, for example leaving a couple of things in highWatermarks on each attempt. I've read through the code, and I don't think this could actually happen right now, but it could after refactoring.

Since we're really just retrying simple API calls, something like this would be safer:
```java
    private <R> R retryUntilSuccessOrThrowOnTaskTimeout(final Supplier<R> supplier,
                                                        final String errorMessage) {
        long deadlineMs = NO_DEADLINE;

        do {
            try {
                return supplier.get();
            } catch (final TimeoutException retryableException) {
                if (taskTimeoutMs == 0L) {
                    throw new StreamsException(
                        String.format(
                            "Retrying is disabled. You can enable it by setting `%s` to a value larger than zero.",
                            StreamsConfig.TASK_TIMEOUT_MS_CONFIG
                        ),
                        retryableException
                    );
                }

                deadlineMs = maybeUpdateDeadlineOrThrow(deadlineMs);

                log.debug(errorMessage, retryableException);
            }
        } while (true);
    }
```

Which would support:
```java
        final Map<TopicPartition, Long> highWatermarks = new HashMap<>(
            retryUntilSuccessOrThrowOnTaskTimeout(
                () -> globalConsumer.endOffsets(topicPartitions),
                String.format(
                    "Failed to get offsets for partitions %s. The broker may be transiently unavailable at the moment.",
                    topicPartitions
                )
            )
        );
```

Incidentally, this seems to be exactly the same as:
```java

        final Map<TopicPartition, Long> highWatermarks = new HashMap<>(
            globalConsumer.endOffsets(topicPartitions, Duration.ofMillis(taskTimeoutMs))
        );
```

right? That might be preferable because we're not implementing any backoff, whereas retrying internally in the client, it can do exponential backoff, etc. (I think).

Reply:
Correct, we rely on the clients internal backoff behavior here. Happy to switch to a `Supplier`